Improve support for viewing PDFs from file://-URLs in the Chrome extension

Rob--W opened the pull request:
Up to now, if a user did not check the "Allow access to file URLs" checkbox at the extension settings page, then they could not view local PDFs in Chromium.

With this patch, such navigations are intercepted, and the extension will ask the user to enable the setting when it detects that it has no access to file:-URLs.

When the user selects "Allow access to file URLs", the extension reloads and Chrome closes all existing extension tabs (https://crbug.com/511670). Since this is disruptive, I've added a work-around (save tab state before page unload, restore tab state when extension loads).

To test:
1. node make chromium
2. chromium --load-extension=build/chromium
3. Visit chrome://extensions and uncheck the 'Allow access to file URLs" at the PDF.js extension (this was automatically checked because you've loaded the extension in unpacked mode).
4. Open a local PDF in the browser.
5. Verify that the PDF.js extension prompts for file access permissions.
6. Follow the instructions:
   - Click on the chrome://extensions link.
   - Click on the "Allow access to file URLs" checkbox at the extensions page.
7. Navigate back
8. Verify that the local PDF is displayed by the extension.<br><br>
9. Go to chrome://extensions in a new tab (so the PDF viewer is still open).
10. Click on "Allow access to file URLs" (to untick the checkbox).
11. Observe that Chrome closes the PDF viewer tab, but verify that the PDF Viewer extension re-opens the tab (and prompts to re-enable file access).

Screenshot:

![Chrome extension asking for access to local files](https://cloud.githubusercontent.com/assets/1365071/8762691/3141a5ac-2d81-11e5-879c-101630f00606.png)

The displayed string adapts itself to the browser. Start Chrome/Chromium with the `LANG=nl_NL.UTF8` environment variable, and you will see a Dutch browser interface and the Dutch translation of "Allow access to file URLs":

![Chrome extension asking for access to local files (LANG=nl_NL.UTF8)](https://cloud.githubusercontent.com/assets/1365071/8762712/7acac154-2d81-11e5-819c-23ba05628e69.png)

Rob--W committed:
[CRX] Do not load local files in non-local frames

Ordinarily, local files cannot be embedded in a non-local website. Until
this commit, the extension allowed websites to embed local PDF files on
non-local (e.g. http(s)) websites. This unintended feature is now
disabled, to align better with Chrome's existing security policies
(=local file:-URLs cannot be loaded in a tab unless expicitly allowed).

Rob--W committed:
Teach users how to enable access to local files.

Rob--W committed:
Add translations of "Allow access to file URLs"

The JSON file is generated as follows.

1. Go to the src/chrome/app/resources directory of Chromium's source.

2. Find the translation ID of the "Allow access to file URLs" string:

grep 'Allow access to file URLs' generated_resources_en-GB.xtb

3. With the ID that you've found, locate the other translations.

grep 3341703758641437857 generated_resources_*.xtb

4. If the result looks OK, serialize the result as JSON and save it.

> path/to/pdf.js/web/chrome-i18n-allow-access-to-file-urls.json \
python -c "import json;print(json.dumps({ \
$(grep 3341703758641437857 generated_resources_*.xtb | \
 sed "s@generated_resources_\([^.]\+\)\.xtb:<translation[^>]\+>\(.\+\)</translation>@'\1':'''\2''',@" \
)}, sort_keys=True, indent=2))"

(Strings are taken from Chromium 45.0.2448.0 (ccrev.com/337313).

timvandermeij commented:
I like the idea here, but I think we should use the current localization facilities to localize the entire message. It seems confusing to get an English message in a Dutch interface with a Dutch part in it. It would be better to just have the entire string in the current localization files such that it will be translated to all other languages upstream. If the "Allow access to file URLs" is a concern (i.e., if the translation should be exactly the same as in Chrome), we should reword the sentence to make it more general, i.e. something like "Allow access to file URLs at chrome://extensions to view this file.". That way we do not mention the exact name of the option, which allows upstream to translate this message more easily and also for languages that Chrome does not even support. What do you think?

Snuffleupagus commented:
In this case, I think that it makes a certain amount of sense to handle the l10n separately.
My main motivation is that it seems slightly strange to essentially require that the Firefox l10n teams translate a Chrome-specific string (since it doesn't appear to be necessary for other builds).

timvandermeij commented:
Okay, that seems reasonable to me. What about the string not being translated entirely, but only partially? Would you agree that that is confusing or do you think it is fine like that?

Rob--W commented:
it would be very nice if the whole string can be translated, but I did not know that it's possible to use Firefox's localization teams to translate the string.

When constructing that sentence, I had the following requirements:

- It should be short and understandable, even to non-technical users.
- It should guide the user to perform the desired action (i.e. enabling file access). This is achieved by inserting a link to chrome://extensions and by showing the extension icon + exact label of checkbox. PDF.js's language is not necessarily equal to Chrome's UI language, so I did not use webL10n to get the correct translation, but some Chrome-specific i18n API to get the UI language.
- It should give context to the action, so that the user knows why he should perform that action: That's why the link to the PDF file is inserted.

That resulted in the final string:

> Click on "[label]" at [link] *line break*
> to view [path to PDF file.]

The line break is inserted to make sure that the top line is always at the same location, regardless of the file path.  

Here's an example for Dutch:

> Druk op "[label]" bij [link] om *line break*
> [path to PDF file.] te weergeven.

Is it really possible to ask Firefox's i18n team to provide such translations?

Rob--W committed:
Restore extension tab after reload

Work-around for https://crbug.com/511670

Rob--W committed:
Don't reload after updating if a tab is in use

Rob--W commented:
I have updated the PR with two more commits, to minimize the disruption caused by reloading the extension, and updated my PR message with 3 more steps-to-verify. Without these two PRs, existing PDF tabs would disappear when the user follows the instructions to enable file access.

Rob--W commented:
@timvandermeij / @Snuffleupagus What prevents this PR from being merged?

timvandermeij commented:
I was wondering if there was a way to translate the English parts of the string, but then I saw that https://github.com/mozilla/pdf.js/blob/master/web/viewer-snippet-mozPrintCallback-polyfill.html#L70 is untranslated as well. Let's add such a TODO to this code too to make it easier to find if we find a way to localize that later on. Personally I'm fine with keeping it this way for the current PR and following up on that later (as it is a TODO just like the other one, someone will find and address it eventually, but it is not a high priority).

Other than that, this just has to be reviewed properly by someone. I'm a bit busy with other PRs and on vacation soon, so could @Snuffleupagus or @yurydelendik review this perhaps?

Rob--W commented:
@timvandermeij I've opened a new ticket (#6333) for the translation issue. Have fun on your well-deserved holiday trip!

Snuffleupagus commented on web/viewer.css:
```diff
@@ -1347,6 +1347,7 @@ html[dir='rtl'] .attachmentsItem > button {
 
 .dialog > .row {
   display: table-row;
+  word-break: break-all;
```
Unconditionally adding this rule leads, in my opinion, to decreased readability of the text in the `DocumentProperties` dialog in some cases; see examples below.
Can you please add this rule only for the `chromeFileAccessOverlay`?

---
 - `intelisa.pdf` from the test-suite

![intelisa](https://cloud.githubusercontent.com/assets/2692120/9277819/d6c7fe6e-42ac-11e5-941c-cfd929d418bd.png)

 - http://mirror.switch.ch/ftp/mirror/tex/macros/latex/contrib/biblatex/doc/biblatex.pdf

![biblatex](https://cloud.githubusercontent.com/assets/2692120/9277924/86c8f25a-42ad-11e5-9a7a-fdd8291a1de9.png)

Rob--W replied:
I'll move the declaration.

( The main reason for adding `word-break: break-all` is to prevent a long sequence of characters (such as a file path) from overflowing. This would be useful for the main dialogs as well, so I put it there. But after testing, I notice that the text doesn't overflow in the document properties dialog. This is because the `p` (containing the text) is not inline (but `table-cell`), has word-wrap `break-word` AND an *explicit* (maximum) width. )

Snuffleupagus commented:
I've tested this and it appears to work exactly as described!
Since this PR is specific to the Chrome extension, I'm treating it somewhat like external code, hence I've not reviewed it line by line.
Once the comment I added is addressed, I think this can be merged.

Snuffleupagus commented:
Thank you for the patch!
